cedar-policy-core/src/evaluator.rs

```diff
@@ -252,9 +252,383 @@ impl<'e> Evaluator<'e> {
+                            format!("operation `{op}` should have integer operands"),
+                        ))
+                    }
+                    // hierarchy membership operator; see note on `BinaryOp::In`
+                    (
+                        Value {
+                            value: ValueKind::Lit(Literal::EntityUID(uid1)),
+                            ..
+                        },
+                        _,
+                        BinaryOp::In,
+                    ) => {
+                        match self.entities.entity(&uid1) {
+                            Dereference::Residual(r) => Ok(PartialValue::Residual(
+                                Expr::binary_app(BinaryOp::In, r, arg2.into()),
+                            )
+                            .try_into()?),
+                            Dereference::NoSuchEntity => self.eval_in(&uid1, None, arg2),
+                            Dereference::Data(entity1) => self.eval_in(&uid1, Some(entity1), arg2),
+                        }
+                    }
+                    (
+                        _,
+                        Value {
+                            value: ValueKind::Set(_),
+                            ..
+                        },
+                        BinaryOp::In,
+                    ) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "`in` is for checking the entity hierarchy;
+                        \
+                        use `.contains()` to test set membership"
+                            .into(),
+                    )),
+                    (
+                        _,
+                        Value {
+                            value: ValueKind::Record(_),
+                            ..
+                        },
+                        BinaryOp::In,
+                    ) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "`in` is for checking the entity hierarchy;
+                        \
+                        use `has` to test if a record has a key"
+                            .into(),
+                    )),
+                    (_, _, BinaryOp::In) => Err(EvaluationError::type_error_with_advice(
+                        nonempty![Type::entity_type(names::ANY_ENTITY_TYPE.clone())],
+                        &arg1,
+                        "the LHS of `in` should be an entity".into(),
+                    )),
+                    // contains, which works on Sets
+                    (_, _, BinaryOp::Contains) => match arg1.value {
+                        ValueKind::Set(Set { fast: Some(h), .. }) => Ok(arg2
+                            .try_as_lit()
+                            .map_or(false, |lit| h.contains(lit))
```

Comment:
maybe would be good to copy the comment from the partial evaluator, that the `map_or(false)` works because we know the set can't contain a non-literal